### ISSUE TITLE
Control log level via configuration

### ISSUE THREAD

**Issue by reubenmiller** [thin-edge/thin-edge.io#1607]
**Is your feature request related to a problem? Please describe.**

It is difficult to activate addition debug information for each of the services.

For example the `tedge_mapper` supports the `tedge_mapper c8y --debug` options, however the only way to activate it is to manually modify the service definition file.

**Describe the solution you'd like**

* Control the activating/deactivating of the `debug` mode of services via a configuration file (ideally through the `tedge.toml` or `system.toml` configuration files). Changing the file will require the service to be restarted
* There should be one configuration item per service so that it can be controlled which components should be doing more detailed logging

**Example configuration**

File: `/etd/tedge/tedge.toml`

```toml
[logging]
tedge_mapper = "info"
tedge_agent = "warn"
tedge_watchdog = "info"
```

**Describe alternatives you've considered**

Editing the service file directly, however this is not easy as the service could be running under a different initd service (e.g. not systemd), which makes process inconsistent for the users.

In the future maybe a non-persistent runtime via an mqtt message would be nice. This would allow a user to change the log level at runtime so they can turn on more detailed logging without restarting the service. This runtime setting would not be persistent, so that in case it would be accidentally left at DEBUG level, then the setting would be reverted after a device or service restart.

**Additional context**

**Comment by PradeepKiruvale** [comment]
- In the existing logging mechanism, the default log levels are info, warn, and error. So, what should be the default log level with the new mechanism?
- Since the new log mechanism is just about setting the log level from the config file (tedge.toml) using the `tedge` cli tool, instead of calling the service binary do we need this to happen dynamically or the effect of new level is only after the restart of the service?

**Comment by reubenmiller** [comment]
1. Though we are looking at changing the underlying logging mechanism, we just want to make the current log level configurable. So I would stick with current default level that we are using.
2. On startup, each service should read the log level currently set in the configuration file. If the configuration file is updated after the start up the service, then the change will only be activated once the service is restarted.

**Comment by albinsuresh** [comment]
> If the configuration file is updated after the start up the service, then the change will only be activated once the service is restarted.

This is gonna be the case 99% of the time, right? I don't think that any customer would start their setup with debug logging enabled right from the start. They'll mostly need at a let point of time when things start going wrong. At that time, expecting the user to manually restart the service defeats the purpose of this feature, I feel.

I'd propose the following alternative:

1. Let the user track the respective systemd definition files by the `c8y-config-plugin`.
1. Let's develop a new `systemd-conifg-plugin` as an extension of the `c8y-config-plugin` which tracks config changes sent by this plugin on `tedge/config-change/<config-type>` topic whenever the tracked systemd definition files change so that it can restart that service (using the commands defined in the `system.toml`). Since this plugin is very generic, the user can use it to reconfigure any systemd service.

**Comment by PradeepKiruvale** [comment]
> 1. Let the user track the respective systemd definition files by the `c8y-config-plugin`.



> > If the configuration file is updated after the start up the service, then the change will only be activated once the service is restarted.
> 
> This is gonna be the case 99% of the time, right? I don't think that any customer would start their setup with debug logging enabled right from the start. They'll mostly need it at a later point of time when things start going wrong. At that time, allowing them to remotely update the config file and then expecting the user to manually restart the service defeats the purpose of this whole feature, right? So I think we need to provide a solution that will do the service restart as well automatically, after the config is updated.
> 
> So, I'd propose the following alternative:
> 
> 1. Let the user track the respective systemd definition files by the `c8y-config-plugin`.
> 2. Let's develop a new `systemd-conifg-plugin` as an extension of the `c8y-config-plugin` which tracks config changes sent by this plugin on `tedge/config-change/<config-type>` topic whenever the tracked systemd definition files change so that it can restart that service (using the commands defined in the `system.toml`). Since this plugin is very generic, the user can use it to reconfigure any systemd service.

This will be helpful only if init system is systemd. What if someone uses some other init system? So, I feel it's not a good idea. This can be a reference implementation like our `watchdog` feature.

**Comment by albinsuresh** [comment]
> This will be helpful only if init system is systemd. What if someone uses some other init system? So, I feel it's not a good idea. This can be a reference implementation like our `watchdog` feature.

The plugin can be written in an init system agnostic way by making it use the control commands defined in the `system.toml` file to start/restart services rather than hardcoding them to systemd.

**Comment by reubenmiller** [comment]
> This is gonna be the case 99% of the time, right? I don't think that any customer would start their setup with debug logging enabled right from the start. They'll mostly need it at a later point of time when things start going wrong. At that time, allowing them to remotely update the config file and then expecting the user to manually restart the service defeats the purpose of this whole feature, right? So I think we need to provide a solution that will do the service restart as well automatically, after the config is updated.

Actually this does not happen too often in practice. Most of the time if you don't have the correct logging level at the time of the incident, dynamically turning it on does not bring you any closer to knowing what happened.

The initial scope was to enable easier initial configuration of the components, so that users did not have to manually edit the service definitions (e.g. systemd, init.d files). And since the proposal is to store the files in `system.toml`, then there is nothing stopping us using the existing configuration file manager plugin.

So I would limit the scope of this ticket from the `enable restart commands in the configuration plugin handler`, and create a new ticket for that functionality.

**Comment by reubenmiller** [comment]
> Let the user track the respective systemd definition files by the c8y-config-plugin.

This is not a good idea, as it would be too easy to break the service definition file (if user's don't know what they are doing), which would could result in the whole thing not working (if someone broke the `tedge-mapper-c8y` or `tedge-mapper-az` service.

The service files are not meant to be edited all the time. So I would much rather treat those files as constants.

**Comment by PradeepKiruvale** [comment]
> > This is gonna be the case 99% of the time, right? I don't think that any customer would start their setup with debug logging enabled right from the start. They'll mostly need it at a later point of time when things start going wrong. At that time, allowing them to remotely update the config file and then expecting the user to manually restart the service defeats the purpose of this whole feature, right? So I think we need to provide a solution that will do the service restart as well automatically, after the config is updated.
> 
> Actually this does not happen too often in practice. Most of the time if you don't have the correct logging level at the time of the incident, dynamically turning it on does not bring you any closer to knowing what happened.
> 
> The initial scope was to enable easier initial configuration of the components, so that users did not have to manually edit the service definitions (e.g. systemd, init.d files). And since the proposal is to store the files in `system.toml`, then there is nothing stopping us using the existing configuration file manager plugin.
> 
> So I would limit the scope of this ticket from the `enable restart commands in the configuration plugin handler`, and create a new ticket for that functionality.

For me this looks like a overkill to have another service to just to monitor and restart a service. Not a good idea for edge devices as there will be less resources.
Another idea is to leverage the `systemd` feature itself to have the same functionality,i.e to reload and restart the service when there is a change with the service definition. May be possible with `systemctl edit service file` when we edit and close the service file, it will auto load and restart the service. Need to explore bit more here if we want to support that dynamic reload/restart option.

**Comment by reubenmiller** [comment]
I would refer back to the original scope of this ticket `Control log level via configuration`. Don't worry about doing anything dynamic now as this will be covered in a future ticket (as it is not a priority at the moment).

Controlling the log level on a by-service basis does add a lot of value (even if it is not dynamic and requires a full device reboot).

**Comment by PradeepKiruvale** [comment]
How to test:

Install the `thin-edge.io` from here https://github.com/thin-edge/thin-edge.io/actions/runs/3639782194

Set the log level in /etc/tedge/system.toml as below.

[log]
tedge_mapper = "trace"
tedge_agent = "info"
tedge_watchdog = "debug"
c8y_log_plugin = "info"
c8y_configuration_plugin = "info"


`sudo tedge connect c8y`

Observe the logs of tedge mapper as below

`journalctl -u tedge-mapper-c8y.service -n 10 --no-pager`
This will print all the log messages, since the log level set to `trace` (Info, debug, warn, error, trace logs)

Observe the logs of tedge agent as below

`journalctl -u tedge-agent.service -n 10 --no-pager`
This will print only `info and warn` log messages, since the log level set to `info`.

Observe the logs of tedge watchdog as below

`journalctl -u tedge-watchdog.service -n 10 --no-pager`
This will print all (info, warn, error, debug)  the log messages, since the log level set to `debug`.

`sudo  c8y_log_plugin`
This will print all (info, warn, error)  the log messages, since the log level set to `info`.

`sudo  c8y_configuration_plugin`
This will print all (info, warn, error)  the log messages, since the log level set to `info`.

**Comment by gligorisaev** [comment]
Checked
